components/vision.py

```diff
@@ -1,213 +1,161 @@
 
-    def __init__(self) -> None:
-        left_rot = Rotation3d(
-            Quaternion(
-                -0.2156163454055786,
-                0.0850898027420044,
-                0.018863987177610397,
-                0.9725808501243591,
-            )
-        )
-        right_rot = Rotation3d(
-            Quaternion(
-                0.21561579406261444,
-                0.08508981764316559,
-                -0.018863940611481667,
-                0.9725809693336487,
-            )
-        )
-        self.cameras = [  # (Camera object, camera-to-robot transform)
-            (
-                PhotonCamera(n),
-                Transform3d(Translation3d(x, y, z), rot).inverse(),
-            )
-            for (n, x, y, z, rot) in [
-                ("cam_port", -0.35001, 0.06583, 0.25, left_rot),
-                ("cam_starboard", -0.35001, -0.06583, 0.247, right_rot),
-            ]
-        ]
-        self.last_timestamps = [0] * len(self.cameras)
-        self.should_log = False
-        # amount of tags that have produced unreasonable estimates in a row
-        self.rejected_in_row = 0.0
-        self.last_z_1 = 0.0
-        self.last_z_2 = 0.0
-
-    def setup(self) -> None:
-        self.field_pos_obj_left = self.field.getObject("vision_pose_left_cam")
-        self.field_pos_obj_right = self.field.getObject("vision_pose_right_cam")
-        self.pose_log_entry = wpiutil.log.FloatArrayLogEntry(
-            self.data_log, "vision_pose"
-        )
+    def on_enable(self) -> None:
+        self.add_to_estimator = True
 
     def execute(self) -> None:
-        for cam_idx, (camera, trans) in enumerate(self.cameras):
-            # if results didnt see any targets
-            if not (results := camera.getLatestResult()).hasTargets():
-                continue
-
-            # if we have already processed these results
-            timestamp = results.getTimestamp()
-            if timestamp == self.last_timestamps[cam_idx] and wpilib.RobotBase.isReal():
+        # if results didn't see any targets
+        if not (results := self.camera.getLatestResult()).hasTargets():
+            return
+
+        # if we have already processed these results
+        timestamp = results.getTimestamp()
+        if timestamp == self.last_timestamp and wpilib.RobotBase.isReal():
+            return
+        self.last_timestamp = timestamp
+
+        # old results cause pose estimator to crash and aren't very useful anyway
+        if abs(wpilib.Timer.getFPGATimestamp() - timestamp) > 0.5:
+            return
+
+        for target in results.getTargets():
+            poses = estimate_poses_from_apriltag(self.camera_to_robot, target)
+            if poses is None:
+                # tag doesn't exist
                 continue
-            self.last_timestamps[cam_idx] = timestamp
+            best_pose, alt_pose, self.last_pose_z = poses
+            pose = choose_pose(
+                best_pose,
+                alt_pose,
```

Comment:
Every use of this poses tuple grabbed each of them by index. Name them for clarity.